Kraken: Fill disruption becomes optional on data load

benoit-bst opened the pull request:
The loading stops no longer when disruption is not loaded.

TODO:
- [ ]  wait for the beginning of a sprint for merge, to let it live... and maybe crash.

benoit-bst committed:
Data : Fill disruption become optionnal

The loading stops no longer when disruption
is not loaded.

benoit-bst commented:
Jenkins, test this please!

benoit-bst committed:
Refactoring : Maintenance worker loop

kinnou02 commented on source/kraken/maintenance_worker.cpp:
```diff
@@ -331,18 +325,12 @@ MaintenanceWorker::MaintenanceWorker(DataManager<type::Data>& data_manager, krak
+    this->load_data();
+
+    // Load Realtime
+    this->load_realtime();
```
`load_realtime` can throw exception.
And from what I can see, one will be throw everytime here.

kinnou02 commented on source/kraken/maintenance_worker.cpp:
```diff
@@ -115,13 +114,7 @@ void MaintenanceWorker::load_realtime(){
-        LOG4CPLUS_ERROR(logger, "Connection to rabbitmq failed: " << ex.what());
-        data_manager.get_data()->is_connected_to_rabbitmq = false;
-        sleep(10);
-    }
```
What is the impact of deleting this?
what change?

kinnou02 commented on source/kraken/maintenance_worker.cpp:
```diff
@@ -331,18 +325,12 @@ MaintenanceWorker::MaintenanceWorker(DataManager<type::Data>& data_manager, krak
         conf(conf),
         next_try_realtime_loading(pt::microsec_clock::universal_time()){
-    try{
-        this->init_rabbitmq();
```
Same thing here, what does it change?

kinnou02 reviewed (changes requested):
With this PR we no longer crash, but we fail silently, that might be worse than crashing.

benoit-bst committed:
Fix rabbitmq init call

benoit-bst committed:
Add tests

benoit-bst committed:
Add disruption status flag

pbougue commented on source/kraken/maintenance_worker.cpp:
```diff
@@ -145,7 +144,8 @@ void MaintenanceWorker::handle_task_in_batch(const std::vector<AmqpClient::Envel
             case pbnavitia::RELOAD:
-                load();
+                this->load_data();
+                this->load_realtime();
```
Not sure, this may fail but it's still unprotected, no?

kinnou02 replied:
errors are handled here https://github.com/benoit-bst/navitia/blob/ade56a81277fc83d536b9215cc3607510382d44e/source/kraken/maintenance_worker.cpp#L128-L132

pbougue commented on source/type/data.cpp:
```diff
@@ -101,22 +104,45 @@ bool Data::load(const std::string& filename,
+    if (chaos_database) {
+        try {
+            fill_disruption_from_database(*chaos_database, *pt_data, *meta, contributors);
+        } catch(const std::exception& ex) {
```
there you should probably reload data (will take a bit of time, but better with clean base_schedule data)
maybe you can be smart : if it fails before doing anything (connection crashing at start) you can avoid reload.

kinnou02 replied:
I agree!

xlqian replied:
@pbougue if you're concerned by the time-consuming reload, we can reload the data(or not) depend on the raised exception type

benoit-bst replied:
I'm agree. I'll do in my next commit.

pbougue commented on source/type/data.cpp:
```diff
@@ -101,22 +104,45 @@ bool Data::load(const std::string& filename,
+
+    // Build Raptor Data
+    try {
+        build_raptor(raptor_cache_size);
```
@kinnou02 if this fails and we loaded disruptions, should we try to reload clean base_schedule data (no disruptions) and see if raptor builds?
Not sure of that, it's a bit of a silent crash... But being resilient is often crashing silently.

kinnou02 replied:
If some disruptions were applied but not all we need to reload the data as you said

xlqian replied:
same idea as above

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,80 @@
+
+  // Load fake data
+  bool check_load = data.load(boost::filesystem::canonical(current_path()).string() \
+  														+ "/" + fake_data_file);
```
weird indent: are you using spaces and not tabs?

kinnou02 replied:
it should be 4 spaces, tabs are evil!

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,80 @@
+
+  // Load fake data
+  bool check_load = data.load(boost::filesystem::canonical(current_path()).string() \
+  														+ "/" + fake_data_file, fake_disruption_path);
```
weird indent

pbougue commented on source/type/data.cpp:
```diff
@@ -101,22 +104,45 @@ bool Data::load(const std::string& filename,
+            fill_disruption_from_database(*chaos_database, *pt_data, *meta, contributors);
+        } catch(const std::exception& ex) {
+            LOG4CPLUS_WARN(logger, "Data disruptions loading failed: " << ex.what());
+            disruption_error = true;
```
There an enum would be more flexible, @kinnou02 ?
Implies new proto PR...

kinnou02 replied:
As I stated when merging the PR proto, a boolean can work for now, but an enum or a free string will be required later on

kinnou02 commented on source/type/data.h:
```diff
@@ -133,6 +133,7 @@ class Data : boost::noncopyable{
     unsigned int version = 0; //< Version of loaded data
     std::atomic<bool> loaded; //< have the data been loaded ?
     std::atomic<bool> loading; //< Is the data being loaded
+    bool disruption_error; // disruption error flag
```
std:atomic

benoit-bst committed:
Add feature : Reloading data

- Reloading data when disruptions failed (Not when Bdd is not connected)
- Add new test for type/data.h
- Update kraken/data_manager.h (Add reloading case)
- Update data_manager_test.cpp

benoit-bst commented:
Jenkins, test this please!

benoit-bst committed:
Fix data_test

using absolute path (Boost)

benoit-bst committed:
Fiw data_test : Compat path

benoit-bst committed:
type/data_set : Fix LC_ALL

benoit-bst committed:
Update data_set

with bug explanation

benoit-bst commented:
Jenkins, test this please!

benoit-bst committed:
Change LC_ALL for data_test

TeXitoi commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,97 @@
+// We create a empty data with lz4 format in current directory.
+// We have to set LC_ALL=C because we can fall on :
+// [Exception] - std::runtime_error: locale::facet::_S_create_c_locale name not valid
+// sources : https://stackoverflow.com/questions/19405272/c-issues-with-boostfilesystem-on-server-localefacet-s-create-c-locale
```
If it fail on tests, wont it also fail on production?

benoit-bst committed:
Update comment in data_set

benoit-bst commented:
Jenkins, test this please!

benoit-bst committed:
data_set : change absolute path (getcwd)

benoit-bst commented:
JIRA :

https://jira.canaltp.fr/browse/NAVITIAII-2359
https://jira.canaltp.fr/browse/NAVITIAII-2360

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -45,8 +45,14 @@ class Data{
             return load_status;
         }
+        bool load_without_disruptions(const std::string&,
+                                     const std::vector<std::string>&,
```
minor: indent

pbougue commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
 }
 
+bool Data::load_without_disruptions(const std::string& filename,
+                                        const std::vector<std::string>& contributors,
```
minor: indent

pbougue commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
+bool Data::load_without_disruptions(const std::string& filename,
+                                        const std::vector<std::string>& contributors,
+                                        const size_t raptor_cache_size) {
+   disruptions_corruption_detected = true;
```
Why not pass it as a parameter and remove the attribute?
Not sure you need it in data_manager (but maybe), and it kinda feels like using the same thing for 2 correlated but different purposes (the code is more complicated to understand):
* one is whether you have to load data -> looks more like a param
* the other is "did it work well?" -> looks more like a return value (or a status attribute like now).

pbougue commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
+        return last_load;
 	}
+
+    // Load Disruption (optionnal)
```
typo: 1 `n` to "optional"

benoit-bst replied:
🙊

benoit-bst committed:
Fix Indent

benoit-bst committed:
spellchecking

kinnou02 commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
+        } catch(const std::exception& ex) {
+            LOG4CPLUS_ERROR(logger, "Data disruptions loading failed: " << ex.what());
+            disruption_error = true;
+            if (std::string(ex.what()) != "Unable to connect to chaos database"){
```
seriously?

kinnou02 commented on source/type/data.h:
```diff
@@ -133,6 +133,8 @@ class Data : boost::noncopyable{
     std::atomic<bool> loaded; //< have the data been loaded ?
     std::atomic<bool> loading; //< Is the data being loaded
+    std::atomic<bool> disruption_error; // disruption error flag
+    bool disruptions_corruption_detected;
```
atomic, always put atomic in Data

kinnou02 commented on source/kraken/data_manager.h:
```diff
@@ -96,6 +96,13 @@ class DataManager{
         success = data->load(database, chaos_database, contributors, raptor_cache_size);
+
+        // If disruptions corruption is detected, reload data without this
+        if ((!success) && (data->disruptions_corruption_detected)) {
```
Looks like responsibilities are mixed between data_manager and Data, only one of them should handle the errors

kinnou02 commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
+            LOG4CPLUS_ERROR(logger, "Data disruptions loading failed: " << ex.what());
+            disruption_error = true;
+            if (std::string(ex.what()) != "Unable to connect to chaos database"){
+                LOG4CPLUS_INFO(logger, "Reload Data without disruptions");
```
It's a warning at least

kinnou02 commented on source/type/data.cpp:
```diff
@@ -101,26 +106,65 @@ bool Data::load(const std::string& filename,
+            }
+        } catch(...) {
+            disruption_error = true;
+            LOG4CPLUS_WARN(logger, "Data disruptions loading failed");
```
and that's a error

benoit-bst committed:
fill_disruption_from_database : Add new catch

benoit-bst committed:
type:data : Add new error handling

benoit-bst committed:
Update data_test.cpp

benoit-bst committed:
Update tests for new type::data methods

benoit-bst committed:
Update data_manager with new type::data methods

Reload data if disruptions failed actived

benoit-bst committed:
Update data_manager_test

benoit-bst committed:
Add new data_manager test

with type::Data

benoit-bst committed:
Compat boost 1.55

benoit-bst committed:
Fix for debian

benoit-bst committed:
Fix for boost

benoit-bst committed:
Change data exceptions location

benoit-bst committed:
Update data_test

benoit-bst committed:
update data_manager_test

benoit-bst committed:
Update docker_test

pbougue commented on source/kraken/fill_disruption_from_database.cpp:
```diff
@@ -47,158 +47,168 @@ void fill_disruption_from_database(const std::string& connection_string,
-                + std::string(e.base().what()));
+    }catch(const pqxx::broken_connection& e){
+        throw;
+        return;
```
Why do you return here?

pbougue commented on source/kraken/fill_disruption_from_database.cpp:
```diff
@@ -47,158 +47,168 @@ void fill_disruption_from_database(const std::string& connection_string,
-        throw navitia::exception(std::string("Unable to connect to chaos database: ")
-                + std::string(e.base().what()));
+    }catch(const pqxx::broken_connection& e){
+        throw;
```
Why catching and throwing right away? Not catching would work the same, no? Maybe you lose part of the stack trace.

pbougue commented on source/kraken/fill_disruption_from_database.cpp:
```diff
@@ -47,158 +47,168 @@ void fill_disruption_from_database(const std::string& connection_string,
-    for (auto ct: count) {
-        impact_count += ct[1].as<int>();
+    } catch(const pqxx::pqxx_exception& e) {
+        throw;
```
Same here (and just below): Why catching and throwing right away? Not catching would work the same, no? Maybe you lose part of the stack trace.

pbougue commented on source/kraken/maintenance_worker.cpp:
```diff
@@ -331,18 +331,25 @@ MaintenanceWorker::MaintenanceWorker(DataManager<type::Data>& data_manager, krak
+    // Load Realtime
     try{
-        load();
+        this->load_realtime();
```
If this raises, shouldn't we cleanup the data and restart with no RT at all?
Maybe partial loading is better than no RT... But raising is probably bad news about the current data state (base+RT), no?
Poke @kinnou02

kinnou02 replied:
It will be safer

pbougue replied:
OK, so we should restart with a clean data (and no RT): same behavior than for chaos db, then.
> :white_check_mark: added to the new JIRA ticket

benoit-bst replied:
I create a new ticket :
https://jira.canaltp.fr/browse/NAVITIAII-2383

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -57,58 +60,64 @@ class Data{
+    // True because fake data don't failed
+    BOOST_CHECK(data_manager.load("fake path"));
+
+    // Load OK, so the internal shared pointer change.
```
typo: pointer change`s`

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -57,58 +60,64 @@ class Data{
+        auto first_data = data_manager.get_data();
+        // True because fake data don't failed
+        BOOST_CHECK(data_manager.load("fake path"));
+        // Load OK, so the internal shared pointer change.
```
typo: pointer change`s`

pbougue commented on source/kraken/tests/CMakeLists.txt:
```diff
@@ -1,9 +1,13 @@
 SET(BOOST_LIBS ${BOOST_LIBS} ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY})
 add_executable(data_manager_test data_manager_test.cpp)
-target_link_libraries(data_manager_test workers utils log4cplus tcmalloc ${Boost_LIBRARIES})
+target_link_libraries(data_manager_test types utils log4cplus tcmalloc ${Boost_LIBRARIES})
```
I am curious about how you tracked the need to change that

benoit-bst replied:
We don´t use workers, but we have to link with type because in data_manager.h, we use type/data_exception.h to handle data errors

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -57,58 +60,64 @@ class Data{
     BOOST_CHECK(data_manager.get_data());
 }
-
-BOOST_AUTO_TEST_SUITE_END()
```
Just curious: How did you track this is not needed?
:white_check_mark: OK, just changed the way boost tests are written, it confused me

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,64 @@ void Data::load(std::istream& ifs) {
+        LOG4CPLUS_WARN(logger, "Unable to connect to disruptions database: " << std::string(ex.what()));
+        disruption_error = true;
+        throw navitia::data::disruptions_broken_connection("Unable to connect to disruptions database: " + std::string(ex.what()));
+    } catch (const pqxx::pqxx_exception& ex){
```
You probably want to log and "log-throw" ex.what() here also ;)

benoit-bst replied:
my mistake :+1:

pbougue replied:
:white_check_mark: Done

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,64 @@ void Data::load(std::istream& ifs) {
+    } catch (const pqxx::broken_connection& ex){
+        LOG4CPLUS_WARN(logger, "Unable to connect to disruptions database: " << std::string(ex.what()));
+        disruption_error = true;
+        throw navitia::data::disruptions_broken_connection("Unable to connect to disruptions database: " + std::string(ex.what()));
```
minor: line length

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,64 @@ void Data::load(std::istream& ifs) {
+    } catch (const pqxx::pqxx_exception& ex){
+        LOG4CPLUS_ERROR(logger, "Disruptions loading error");
+        throw navitia::data::disruptions_loading_error("Disruptions loading error");
+        disruption_error = true;
```
This should be moved up (unreachable)

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,64 @@ void Data::load(std::istream& ifs) {
+    } catch (const std::exception& ex) {
+        LOG4CPLUS_ERROR(logger, "Disruptions loading error: " << std::string(ex.what()));
+        throw navitia::data::disruptions_loading_error("Disruptions loading error: " + std::string(ex.what()));
+        disruption_error = true;
```
This should be moved up (unreachable)

benoit-bst replied:
:white_check_mark:  Done

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,64 @@ void Data::load(std::istream& ifs) {
+    } catch (...) {
+        LOG4CPLUS_ERROR(logger, "Disruptions loading error");
+        throw navitia::data::disruptions_loading_error("Disruptions loading error");
+        disruption_error = true;
```
This should be moved up (unreachable)

benoit-bst replied:
:white_check_mark: Done

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    data.save(fake_data_file);           \
+
+// Absolute path
+std::string complete_path(const std::string file_name){
```
This function should be `static`, no?
Also shouldn't it be in utils or utils_test as it's duplicated? Don't you have it in boost file system directly?

benoit-bst replied:
The equivalent of complete_path exist in Boost but there is a bug with boost < 1.56 failed.
We have to set LC_ALL=C because we can fall on :
[Exception] - std::runtime_error: locale::facet::_S_create_c_locale name not valid
When you create a absolute path, boost < 1.56 failed. 
sources : https://stackoverflow.com/questions/19405272/c-issues-with-boostfilesystem-on-server-localefacet-s-create-c-locale

So this is more easy for compatibility to use complete_path (C style)

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    data.save(fake_data_file);           \
+
+// Absolute path
+std::string complete_path(const std::string file_name){
```
This is duplicate code (ctrl + F "complete_path"), same comment applies.

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    BOOST_CHECK_EQUAL(data.last_load, false);
+    try {
+        data.load_nav(complete_path(fake_data_file));
+    } catch(const navitia::data::data_loading_error& ex) {
```
As you don't use `ex` would be better not naming it, to make it explicit and avoid warnings.

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    BOOST_CHECK_EQUAL(data.last_load, true);
+    try {
+        data.load_nav("wrong_path");
+    } catch(const navitia::data::data_loading_error& ex) {
```
Same: as you don't use `ex` would be better not naming it, to make it explicit and avoid warnings.

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    bool failed = false;
+    try {
+        data.load_nav("wrong_path");
+    } catch(const navitia::data::data_loading_error& ex) {
```
Same: as you don't use `ex` would be better not naming it, to make it explicit and avoid warnings.

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    bool failed = false;
+    try {
+        data.load_disruptions(fake_disruption_path);
+    } catch(const navitia::data::disruptions_broken_connection& ex) {
```
Same: as you don't use `ex` would be better not naming it, to make it explicit and avoid warnings.

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+        // load .nav.lz4
+        try {
+            data->load_nav(filename);
+        } catch(const navitia::data::data_loading_error& ex) {
```
`ex` must be logged (probably in ERROR)

benoit-bst replied:
It is already logged in type/data.cpp

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+            try {
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection& ex){
+
```
A comment would be nice to explain why we're doing nothing, also a log explaining we're going without disruptions.

benoit-bst replied:
:+1:

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+            bool load_disruptions_failed = false;
+            try {
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection& ex){
```
`ex` must be logged (probably in ERROR)

benoit-bst replied:
It is already logged in type/data.cpp

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection& ex){
+
+            } catch(const navitia::data::disruptions_loading_error& ex) {
```
`ex` must be logged (probably in ERROR)

benoit-bst replied:
It is already logged in type/data.cpp

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+                try {
+                    data = create_data(data_identifier.load());
+                    //data->load_nav(filename);
+                } catch(const navitia::data::data_loading_error& ex) {
```
`ex` must be logged (probably in ERROR)

benoit-bst replied:
It is already logged in type/data.cpp

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+        // Build Raptor Data
+        try {
+            data->build_raptor(raptor_cache_size);
+        } catch(const navitia::data::raptor_building_error& ex) {
```
`ex` must be logged (probably in ERROR)

benoit-bst replied:
It is already logged in type/data.cpp

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -1,52 +1,55 @@
-            return load_status;
-        }
+
+        void load_nav(const std::string& filename){}
```
As you don't use the param, please don't name it: less warning, more explicit

benoit-bst replied:
:+1:

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -1,52 +1,55 @@
-        }
+
+        void load_nav(const std::string& filename){}
+        void load_disruptions(const std::string& database,
```
Same here for the 2 params: as you don't use them please don't name: less warning, more explicit

pbougue commented on source/kraken/tests/data_manager_test.cpp:
```diff
@@ -1,52 +1,55 @@
+        void load_nav(const std::string& filename){}
+        void load_disruptions(const std::string& database,
+                              const std::vector<std::string>& contributors = {}){}
+        void build_raptor(size_t cache_size = 10){}
```
Same (copy/paste syndrome :smiley:): as you don't use the param, please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -182,9 +182,18 @@ BOOST_FIXTURE_TEST_CASE(fusio_test, ArgsFixture) {
+    bool failed = false;
+    try {
+        data.load_nav(input_file);
+    } catch(const navitia::data::data_loading_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -182,9 +182,18 @@ BOOST_FIXTURE_TEST_CASE(fusio_test, ArgsFixture) {
+    }
+    try {
+        data.build_raptor();
+    } catch(const navitia::data::raptor_building_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -196,9 +205,19 @@ BOOST_FIXTURE_TEST_CASE(fusio_test, ArgsFixture) {
+    bool failed = false;
+    try {
+        data.load_nav(input_file);
+    } catch(const navitia::data::data_loading_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -196,9 +205,19 @@ BOOST_FIXTURE_TEST_CASE(fusio_test, ArgsFixture) {
+    }
+    try {
+        data.build_raptor();
+    } catch(const navitia::data::raptor_building_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -345,9 +364,18 @@ BOOST_FIXTURE_TEST_CASE(ntfs_v5_test, ArgsFixture) {
+    bool failed = false;
+    try {
+        data.load_nav(input_file);
+    } catch(const navitia::data::data_loading_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/ed/docker_tests/ed_integration_tests.cpp:
```diff
@@ -345,9 +364,18 @@ BOOST_FIXTURE_TEST_CASE(ntfs_v5_test, ArgsFixture) {
+    }
+    try {
+        data.build_raptor();
+    } catch(const navitia::data::raptor_building_error& ex) {
```
As you don't use `ex` please don't name it: less warning, more explicit

pbougue commented on source/routing/routing_cli_utils.cpp:
```diff
@@ -49,9 +49,9 @@ namespace navitia { namespace cli {
         void compute_options::load(const std::string& file) {
```
OK for those changes, but is this a :mushroom: place?
@kinnou02 @TeXitoi this looks like dead code / never used exe, no? :smile:

kinnou02 replied:
I think it's :mushroom: place!

pbougue replied:
This is probably dead code, so to be removed... But we have to look thoroughly into that, so that's not the matter now, will be done after.

pbougue commented on source/type/data.h:
```diff
@@ -186,7 +180,7 @@ class Data : boost::noncopyable{
 
 
-    bool last_load = true;
+    bool last_load;
```
I know it was there before, but naming is not great.
Are you even sure we need this and it's not redundant with `std::atomic<bool> loaded` ?
If we need this, both need a better naming, or to be moved, I guess.
Using "is" in naming bool is more understandable.

pbougue commented on source/type/data_exceptions.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "data_exceptions.h"
```
:+1:

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+                LOG4CPLUS_ERROR(logger, "Reload data: " << filename);
+                try {
+                    data = create_data(data_identifier.load());
+                    //data->load_nav(filename);
```
Here you have a problem.
Pretty much all the mechanism should be the same as few lines above. It probably deserves a method to mutualize.
Then it might be called directly in the catch (not sure).

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+            data->build_raptor(raptor_cache_size);
+        } catch(const navitia::data::raptor_building_error& ex) {
+            data->loading = false;
+            return false;
```
Can't we try without disruptions if we had some, before rage-quit? :smiley: 
Poke @kinnou02

kinnou02 replied:
Not sure about `build_raptor` failing, there is no explicit throw in it.
For now let just abort loading if it fail

benoit-bst replied:
I just checked, there is no throw in Raptor data building. I removed the error handling for the moment

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    // real data : Loading successed
+    BOOST_CHECK(data_manager.load(complete_path(fake_data_file)));
+
+    // Load success, so the internal shared pointer change.
```
typo (same multiple times below): "change`s`"

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+// We create a empty data with lz4 format in current directory.
+#define CREATE_FAKE_DATA(fake_file_name) \
+    navitia::type::Data data(0);         \
+    data.save(fake_data_file);           \
```
Can't we use tmp directory instead?
Not fan of tests creating files... Or we have to clean after.

benoit-bst replied:
Yes, i can remove empty .lz4 after.

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    // Fake data : Loading failed
+    BOOST_CHECK(!data_manager.load("wrong path"));
+
+    // Load failed, so the internal shared pointer no change.
```
English: `does not` change

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    // Load failed, so the internal shared pointer no change.
+    // Data has not changed.
+    auto second_data = data_manager.get_data();
```
Minor: No need to store the result.

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+static const std::string fake_disruption_path = "fake_disruption_path";
+
+// We create a empty data with lz4 format in current directory.
+#define CREATE_FAKE_DATA(fake_file_name) \
```
Same: duplicate, and no file cleanup after.

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+static const std::string fake_disruption_path = "fake_disruption_path";
+
+// We create a empty data with lz4 format in current directory.
+#define CREATE_FAKE_DATA(fake_file_name) \
```
I do not get why it has to be a macro.

benoit-bst replied:
I change it in function style (better for debug)

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,113 @@
+    // load .nav
+    bool failed = false;
+    try {
+        data.load_nav("wrong_path");
```
Redundant with the end of the previous test, no?

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    BOOST_CHECK_EQUAL(third_data, fourth_data);
+}
+
+BOOST_AUTO_TEST_CASE(load_failed) {
```
Isn't that teste redundant with the end of the previous one?
Also, you did not test for all the branches of the load:
:white_check_mark:  only nav, and nav fail are done
:question: missing the parts where we smartly reload (or not)

benoit-bst replied:
To test reloading, we have to create a disruption bdd and fail during the loading. I think it´s complicated in a Unit Test.

pbougue reviewed (changes requested):
Part of the comments are from clang warnings.
Don't know about how easy are the tests on smart reload, but you missed some commented part of your code without them.

kinnou02 commented on source/type/data_exceptions.h:
```diff
@@ -0,0 +1,77 @@
+};
+
+// Data loading exceptions handler
+struct data_loading_error : public navitia::exception {
```
You might want to inherit from `recoverable_exception` here

benoit-bst replied:
:+1:

kinnou02 commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
+                LOG4CPLUS_ERROR(logger, "Reload data: " << filename);
+                try {
+                    data = create_data(data_identifier.load());
+                    //data->load_nav(filename);
```
why this is commented?

benoit-bst replied:
my mistake

kinnou02 commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,65 @@ class DataManager{
-        success = data->load(database, chaos_database, contributors, raptor_cache_size);
-        if (success) {
-            set_data(std::move(data));
+        data->loading = true;
```
This isn't really useful, but if you want to keep it you have to use RAII or finally to set it back to false

pbougue replied:
This comment is still valid.

benoit-bst replied:
we have 3 cases :
- load .nav.lz4
   we catch all errors types (in type/data.cpp we pass all). Inside the catc,h we pass data->loading = false
   So this is safe :white_check_mark: 
- load disruptions
   we catch all errors types (in type/data.cpp we pass all). Inside the catch we pass data->loading = false
   So this is safe :white_check_mark: 
- Build raptor
   the function is unprotected 
```C++
// if building failed data->loading == true
data->build_raptor(raptor_cache_size);
data->loading = false;
``` 
But when the build_raptor failed, kraken is dead no ? So, this is not important

benoit-bst committed:
Update ed_integration_tests.cpp

data.build_raptor() has no error handling

benoit-bst committed:
Update data_mamager.h

Remove handling error with data.build_raptor()

benoit-bst committed:
fill_disruption_from_database.cpp

Remove try-catch, the errors handling is up

benoit-bst committed:
data_manager_test.cpp

- Fix warning
- Fix language errors

benoit-bst committed:
data_manager_with_type_data_test.cpp

- Clean syntax
- Remove lz4 test file

benoit-bst committed:
worker.cpp

last_load -> last_load_succeeded

benoit-bst committed:
type/data : update errors handling

benoit-bst committed:
data_test.cpp

- Remove duplicate tests
- Fix warning

benoit-bst committed:
Fix warning + remove fake file in data_tests

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,56 @@ class DataManager{
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection&){
+                LOG4CPLUS_WARN(logger, "Load data without disruptions");
+
```
Useless extra line.
Instead a comment on why you don't need to reload (data are clean, nothing was applied) would be nice.

benoit-bst replied:
:white_check_mark: Done

benoit-bst replied:
I add a comment to explain more precisely.

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,56 @@ class DataManager{
+            try {
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection&){
+                LOG4CPLUS_WARN(logger, "Load data without disruptions");
```
Log should be more explicit about "broken connection to db"

benoit-bst replied:
There is a log inside data class, it gives that:

```bash
[18-02-22 09:46:21,214] data.cpp:152 [WARN ] - Unable to connect to disruptions database: missing "=" after "fake_name_path" in connection info string 
[18-02-22 09:46:21,222] data_manager.h:120 [WARN ] - Load data without disruptions 
```

benoit-bst replied:
:white_check_mark: Done

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,56 @@ class DataManager{
+                // Reload data .nav.lz4
+                LOG4CPLUS_ERROR(logger, "Reload data with disruptions: " << filename);
+                try {
+                    data = create_data(data_identifier.load());
```
Now it's OK (uncommented) but it's still redundant code, so a method would be nice (for later corrections/changes)

benoit-bst replied:
:white_check_mark:  Done

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,56 @@ class DataManager{
+        if (chaos_database != boost::none) {
+            try {
+                data->load_disruptions(*chaos_database, contributors);
+            } catch (const navitia::data::disruptions_broken_connection&){
```
minor: spacing `catch(...) {`

benoit-bst replied:
:white_check_mark:  Done

benoit-bst committed:
Update data_manager_test.cpp (old way)

pbougue commented on source/kraken/data_manager.h:
```diff
@@ -88,17 +89,56 @@ class DataManager{
-        return success;
+
+        // Build Raptor Data
+        data->build_raptor(raptor_cache_size);
```
Poke @kinnou02 not sure if catching and returning false here could have an interest.

benoit-bst replied:
It is use here :
https://github.com/CanalTP/navitia/blob/dev/source/kraken/maintenance_worker.cpp#L60

benoit-bst committed:
data_manager_with_type_data_test.cpp

remove macro function

benoit-bst committed:
data_exception.h

inherit from recoverable_exception

pbougue commented on source/type/data.cpp:
```diff
@@ -62,11 +61,10 @@ namespace pt = boost::posix_time;
 
-wrong_version::~wrong_version() noexcept {}
-
 const unsigned int Data::data_version = 67; //< *INCREMENT* every time serialized data are modified
```
You have to increment that number when you change data serialization (so that kraken knows and crash cleanly when the nav file is not in the good version)

benoit-bst replied:
???       I can´t change it, because it is a static const declaration no ?
https://github.com/CanalTP/navitia/blob/dev/source/type/data.h#L132

pbougue replied:
You're probably right about not needing to increment as type did not change...
But we want to be safe, so please increment :)

benoit-bst replied:
Okay for safety, I increment the data
:white_check_mark:

pbougue commented on source/type/data.h:
```diff
@@ -186,7 +180,7 @@ class Data : boost::noncopyable{
 
 
-    bool last_load = true;
+    bool last_load_succeeded;
```
So is the use really different from `loaded`?
It seems to me that they work the same (didn't look closely, though).

benoit-bst replied:
I think you have right, but this decision can change a lot of code, no ?
https://github.com/CanalTP/navitia/blob/dev/source/type/data.h#L218

benoit-bst replied:
poke @kinnou02

benoit-bst replied:
I create an other ticket for this feature :
https://jira.canaltp.fr/browse/NAVITIAII-2383

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,126 @@
+static const std::string fake_disruption_path = "fake_disruption_path";
+
+// We create a empty data with lz4 format in current directory.
+static void create_fake_data(const std::string& fake_file_name){
```
This still needs to be shared code (and so should "complete_path()"), in a `utils_test.h` probably https://github.com/CanalTP/navitia/blob/dev/source/tests/utils_test.h.
That way you will also be able to reuse it in data_test.cpp

benoit-bst replied:
Currently, utils_test.h is not a library. Furthermore, functions in utils_test.h are only use in local directory (sources/tests). This is a local helper file.
* So if you want to create a shared library for tests it is possible, but you have to define where and  what's inside.
* An other possibility is to include the shared function in utils

Maybe the cost development is too much just for that, no ?

benoit-bst replied:
As agreed, i will move complete_path inside utils/functions because it is used in kraken_zmq.cpp and tests too

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+// Absolute path
+static std::string complete_path(const std::string file_name){
```
Duplicate code (to be shared)

pbougue replied:
And thanks for the explanation on boost bug! This should go in a comment, so someone doesn't try to refactor using boost :wink:

benoit-bst replied:
:+1:

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,126 @@
+
+BOOST_AUTO_TEST_CASE(load_failed) {
+
+    create_fake_data(fake_data_file);
```
You forgot to remove the created file.

benoit-bst replied:
my mistake,
:white_check_mark:  Done

pbougue commented on source/type/data.cpp:
```diff
@@ -129,6 +132,55 @@ void Data::load(std::istream& ifs) {
+    } catch (const pqxx::pqxx_exception& ex){
+        LOG4CPLUS_ERROR(logger, "Disruptions loading error: " << std::string(ex.base().what()));
+        throw navitia::data::disruptions_loading_error("Disruptions loading error: " + std::string(ex.base().what()));
+        disruption_error = true;
```
Still unreachable (same below) :wink:

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,101 @@
+
+BOOST_AUTO_TEST_CASE(load_data) {
+
+    navitia::type::Data data(0);
```
To be change for the use of shared function

benoit-bst replied:
move in utils

pbougue commented on source/type/tests/data_test.cpp:
```diff
@@ -0,0 +1,101 @@
+static const std::string fake_disruption_path = "fake_disruption_path";
+
+// Absolute path
+static std::string complete_path(const std::string file_name){
```
Duplicate code (to be shared)

benoit-bst replied:
move in utils

pbougue commented on source/kraken/tests/data_manager_with_type_data_test.cpp:
```diff
@@ -0,0 +1,126 @@
+    // Fake data : Loading failed
+    BOOST_CHECK(!data_manager.load("wrong path"));
+
+    // Load failed, so the internal shared pointer no changes.
```
`does not` change

benoit-bst replied:
:white_check_mark:  Done

pbougue reviewed (commented):
Hope it's not too messy...
I reported some comments from previous review, as they where not entirely managed from my point of view.

benoit-bst committed:
Update utils submodule

add navitia::absolute_path function

benoit-bst committed:
Update data_manager

Add private function to avoid code redundancy

benoit-bst committed:
Update code with utils submodule

add absolute_path function

pbougue reviewed (approved):
All requested changes are done or tracked in another JIRA ticket... Looks good.

benoit-bst committed:
Increment data version : 67 -> 68
